### PR TITLE
feat: add catalog search, sort, pagination

### LINKED ISSUE #135: Catalog search, sort, filter, pagination
- [x] Sort alphabetically, by status
- [x] Update query string when sorting
- [x] Persist sort to local storage
- [x] View by list or grid
- [x] Update query string when changing view
- [x] Persist to local storage when changing view
- [x] Update query string when searching
- [x] Search catalog items with in-memory data based on name and description
- [x] Pagination functionality with in-memory data
- [x] Update query string when updating pagination
- [x] Sticky scroll
- [x] No results empty state

### PR DESCRIPTION
**Pull request by mattrosno**
Closes #135

Makes catalog controls interactive.

#### Changelog

**New**

- Search functionality, persists to query string
- Sort functionality, persists to query string, local storage per view
- View functionality, persists to query string, to local storage per view
- Pagination functionality, persists to query string
- Empty state (no results found)
- `useEventListener` hook
- `useQueryState` hook to derive state from URL query, including parser and serializer for string, numbers, boolean, and dates, as well as default value functionality and persisting to local storage

**Changed**

- `media-query.js` file name to be consistent with the `use-` naming convention for hooks

**Removed**

- Data print-outs on catalog pages

#### Testing / reviewing

N/A

### CODE REVIEW

**Review by andreancardona — DISMISSED** [review]
@mattrosno this looks so awesome! 🙌🏽

File: services/web-app/components/catalog-sort/catalog-sort.js [hunk 1]
```diff
@@ -6,65 +6,95 @@
...
+  const scrollHandler = useCallback(() => {
+    if (containerRef.current && typeof window !== 'undefined') {
+      setIsSticky(containerRef.current.getBoundingClientRect().top === 112)
+    }
+  }, [])
+
+  useEventListener('scroll', scrollHandler)
+  useEffect(() => {
+    scrollHandler()
+  }, [scrollHandler])
```
Comment (jdharvey-ibm) [inline services/web-app/components/catalog-sort/catalog-sort.js:45]: Depending on what use cases you envision for `useEventListener`, you could potentially put the `useCallback` inside of the `useEventListener` hook to remove the need for each hook user to memoize their callback before providing it as an arg.

File: services/web-app/utils/use-query-state.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  // When not in a SSR context, get the query value directly from URLSearchParams so there's no
+  // reliance on router.query which may not be populated yet.
+  const getValue = useCallback(() => {
+    if (typeof window === 'undefined') {
+      return null
+    }
+
+    const query = new URLSearchParams(window.location.search)
+    const queryValue = query.get(key)
+    const storageValue = useStorage ? localStorage.getItem(`${router.pathname}:${key}`) : null
```
Comment (jdharvey-ibm) [inline services/web-app/utils/use-query-state.js:58]: Future us TODO:
Based on @jeanservaas decision on 11/2, filters should persist across a session, but when the session ends/a new one starts, the filters should all reset to the global defaults or team defaults.

File: services/web-app/utils/use-query-state.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  }
+}
+
+export const useQueryState = (
+  key,
+  {
+    defaultValue = '',
+    parse = queryTypes.string.parse,
+    serialize = queryTypes.string.serialize,
+    useStorage = false
```
Comment (jdharvey-ibm) [inline services/web-app/utils/use-query-state.js:44]: minor: naming convention on this boolean makes it look like a react hook. Maybe something like `shouldUseStorage` instead?

File: services/web-app/utils/use-query-state.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+      if (newValue === null || newValue === undefined) {
+        query.delete(key)
+      } else {
+        query.set(key, serialize(newValue))
+      }
+
+      router.replace(`?${query.toString()}`, undefined, { shallow: true, scroll: false })
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [getValue, key, serialize]
```
Comment (jdharvey-ibm) [inline services/web-app/utils/use-query-state.js:85]: Which dep in this block does the linter not like?

File: services/web-app/utils/use-query-state.js (hunk 3 above)
Comment (mattrosno) [reply services/web-app/utils/use-query-state.js:44]: I'll update to `saveToStorage`.

File: services/web-app/utils/use-query-state.js (hunk 4 above)
Comment (mattrosno) [reply services/web-app/utils/use-query-state.js:85]: That `useCallback` calls `router.replace()` above it two lines, which then triggers unnecessary renders for some reason. The Next.js router must update that function for some reason.

File: services/web-app/utils/use-query-state.js (hunk 2 above)
Comment (mattrosno) [reply services/web-app/utils/use-query-state.js:58]: Okay, we can look into this later. I'm persisting sort and view, not the filter yet. I'd say I'm persisting view preferences so far, not any filters which reduce what is shown in the catalogs.

File: services/web-app/components/catalog-sort/catalog-sort.js (hunk 1 above)
Comment (mattrosno) [reply services/web-app/components/catalog-sort/catalog-sort.js:45]: That gives an eslint warning in the static code analysis. It cool if I keep it as is?

image

File: services/web-app/utils/use-query-state.js (hunk 4 above)
Comment (jdharvey-ibm) [reply services/web-app/utils/use-query-state.js:85]:
```suggestion
    // The Next.js router updates `router.replace`, which should not trigger re-renders
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [getValue, key, serialize]
```

File: services/web-app/components/catalog-sort/catalog-sort.js (hunk 1 above)
Comment (jdharvey-ibm) [reply services/web-app/components/catalog-sort/catalog-sort.js:45]: Ohhh. Sorry. I didn't realize it was getting used below in the `useEffect` hook as well.

Comment (mattrosno) [reply services/web-app/components/catalog-sort/catalog-sort.js:45]: Ya! Just to make sure the handler fires once on page load too.

**Review by andreancardona — APPROVED** [review]

**Pull request merged**